Implement client protocol v2

AlekseiS opened the pull request:
Please see the following wiki page for v2 protocol description and key differences compared to v1: https://github.com/prestodb/presto/wiki/Presto-v2-client-protocol

Limitations of the current implementation:
* Client v2 only supports one download channel.
* Information about desired parallelism needs to be passed from a client. Parallelism of the last stage will be used.
* Information about desired parallelism needs to be taken into account by planner to add an arbitrary buffer on top. Currently, it works without additional buffers.
* Information about desired parallelism needs to be taken into account by scheduler to schedule the right number of arbitrary buffer tasks.
* Information about using coordinator only download needs to be passed from the client and used by planner/scheduler to fulfil it.
* Certain DDL query shapes need to output "true" as a result of their operation.
* jdbc needs to be able to select v2 version. Currently, it can only be done in CLI.
* Json representation of structural types is the same as in v1. There were some thoughs about representing it differently.
* "Actions", such as "set transaction id", "clear session properties", etc are represented the same way as in v1.
* Query is started immediately when submitted. There's a consideration to start a query on the first status request.
* Need to clarify/change how data uris are being sent to a client. Currently, they will be sent on every update in a cumulative way, but only once each time.

kokosing commented on presto-main/src/main/java/com/facebook/presto/server/protocol/Query.java:
```diff
@@ -60,184 +51,99 @@
     }
 
-    private Query(
+    public static Query createV2(
```
Why not to extract `QueryFactory`? Then it would be nice to have an interface of  `Query` defined and if needed `AbstractQuery` as well.

AlekseiS replied:
Well, createV1 and createV2 need different parameters. Public methods of `Query` abstract class is pretty much the interface. In any case, let me think a bit more about it.

kokosing commented on presto-main/src/main/java/com/facebook/presto/server/protocol/StatementResourceV2.java:
```diff
@@ -0,0 +1,167 @@
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+@Path("/v2/statement")
+public class StatementResourceV2
```
I do not see `StatementResouceV1` defined. It would be nice to follow the same convention as for `Query`.

AlekseiS replied:
I didn't rename it mainly to make it easier to rebase and modify this PR. I suggest renaming it once this PR is merged.

kokosing commented on presto-client/src/main/java/com/facebook/presto/client/StatementClientV2.java:
```diff
@@ -0,0 +1,633 @@
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+@ThreadSafe
+public class StatementClientV2
```
ah... here it comes...

kokosing commented on presto-client/src/main/java/com/facebook/presto/client/StatementClientFactory.java:
```diff
@@ -15,12 +15,29 @@
+
 public final class StatementClientFactory
 {
+    public static final ProtocolVersion DEFAULT_PROTOCOL_VERSION = ProtocolVersion.V1;
```
I would remove this static field, instead I would inject here protocol version to factory construct from configuration/cli parameters and create an instance of factory that would be used multiple times. What do you think?

AlekseiS replied:
Both v1 and v2 might be used by different statement clients in the same process, for example unit tests use it. As a result, it cannot be a singleton initialized at start up. What class/component will then own an instance of the version-specific factory?

findepi commented:
@AlekseiS @kokosing is there a documentation how protocol v2 differs from v1?
(i vaguely remember seeing something like that but I wouldn't assume it was completed or it matches implementation in 100%)

sopel39 commented:
@AlekseiS I am also curious is there a plan to add support for very large data (e.g: BLOBs), e.g: tens of MBs? Would be super useful if we wanted to store compiled methods for execution during query (@dain ?)

I know that worker to worker has page size limitation right now, but I would guess client protocol is another story.

AlekseiS commented:
Rebased

AlekseiS commented:
@sopel39  Do you want to support large data as part of protocol control information, like properties, prepared statements, etc? Or do you mean downloading large amounts of data by the client?

AlekseiS commented:
Updated PR description to include a wiki link to protocol description and key differences with v1. Also, mention limitations of the current implementations.

AlekseiS commented on presto-benchmark-driver/src/main/java/com/facebook/presto/benchmark/driver/BenchmarkQueryRunner.java:
```diff
@@ -151,18 +151,51 @@ public BenchmarkQueryResult execute(Suite suite, ClientSession session, Benchmar
         while (true) {
             // start query
-            StatementClient client = newStatementClient(okHttpClient, session, "show schemas");
+            try (StatementClient client = newStatementClient(okHttpClient, session, "show schemas")) {
```
I see you already have a viable approach in your PR. I'll rebase on it once it's merged.

AlekseiS commented on presto-benchmark-driver/src/main/java/com/facebook/presto/benchmark/driver/BenchmarkQueryRunner.java:
```diff
@@ -151,18 +151,51 @@ public BenchmarkQueryResult execute(Suite suite, ClientSession session, Benchmar
+                // read query output
+                ImmutableList.Builder<String> schemas = ImmutableList.builder();
+                while (client.isValid() && client.advance()) {
+                    // we do not process the output
```
Probably not. I can remove it, but it's not related to my changes.

AlekseiS commented on presto-cli/src/main/java/com/facebook/presto/cli/QueryRunner.java:
```diff
@@ -68,7 +68,7 @@ public QueryRunner(
         OkHttpClient.Builder builder = new OkHttpClient.Builder();
 
-        setupTimeouts(builder, 5, SECONDS);
+        setupTimeouts(builder, 30, SECONDS);
```
Yes, it's going to affect both. Commit message just gives v2 as an explanation why. I think it shouldn't be a problem to increase it for both. However, if someone thinks otherwise I can configure different timeouts based on protocol version.

AlekseiS commented on presto-client/src/main/java/com/facebook/presto/client/QueryActions.java:
```diff
@@ -0,0 +1,164 @@
+    @Override
+    public String toString()
+    {
+        return toStringHelper(this)
```
Done

AlekseiS commented on presto-client/src/main/java/com/facebook/presto/client/QueryActions.java:
```diff
@@ -0,0 +1,164 @@
+    }
+
+    @Nullable
+    public static QueryActions createIfNecessary(
```
So, that it can directly be passed as a field value in `QueryResults`. Also, it seems to be more inline with the current style of using nullable fields for classes used as json structures.

AlekseiS commented on presto-client/src/main/java/com/facebook/presto/client/StatementClientFactory.java:
```diff
@@ -0,0 +1,26 @@
+
+import okhttp3.OkHttpClient;
+
+public final class StatementClientFactory
```
I need to think about it. One thing I noticed is that CLI creates a new http client for every query by copying properties from existing one.  See `QueryRunner.startInternalQuery()`, line 118. @electrum , is it intentional?
If so then factory won't be able to own `httpClient`.

AlekseiS commented on presto-jdbc/src/main/java/com/facebook/presto/jdbc/QueryExecutor.java:
```diff
@@ -42,7 +43,7 @@ public QueryExecutor(OkHttpClient httpClient)
+        return newStatementClient(httpClient, session, query);
     }
 
     public ServerInfo getServerInfo(URI server)
```
I think you're suggesting wider changes to how clients are managed across CLI, jdbc, etc. It might be possible, but I'd rather do it as a different PR.

AlekseiS commented on presto-main/src/main/java/com/facebook/presto/server/SessionContext.java:
```diff
@@ -16,38 +16,55 @@
     String getSchema();
 
+    @Nullable
     String getSource();
```
In general, yes, but somehow it was done this way. I added annotations to at least make it clear that some of these fields are nullable.

electrum replied:
Yes, it would be nice to convert these to optionals, but that's a bigger project.

AlekseiS commented on presto-main/src/main/java/com/facebook/presto/server/SessionContext.java:
```diff
@@ -16,38 +16,55 @@
+    @Nullable
     String getSource();
 
+    @NotNull
```
Removed.

AlekseiS commented on presto-main/src/main/java/com/facebook/presto/server/TaskResource.java:
```diff
@@ -241,19 +248,7 @@ public void getResults(@PathParam("taskId") TaskId taskId,
             @Suspended AsyncResponse asyncResponse)
             throws InterruptedException
     {
-        requireNonNull(taskId, "taskId is null");
```
Good idea. I moved method extraction in this class to a separate commit.

AlekseiS commented on presto-main/src/main/java/com/facebook/presto/server/TaskResource.java:
```diff
@@ -326,4 +345,98 @@ private static Duration randomizeWaitTime(Duration waitTime)
+        }
+
+        @GET
+        @Path("{taskId}/results/{bufferId}/{token}")
```
It's new in this class, because the logic is different: convert to json vs convert to pages. It uses the same approach with `RowIterable` as `StatementClient`.

AlekseiS commented on presto-tests/src/main/java/com/facebook/presto/tests/DistributedQueryRunner.java:
```diff
@@ -432,4 +427,50 @@ private static void closeUnchecked(AutoCloseable closeable)
+            this.workerCount = workersCount;
+        }
+
+        public Builder setExtraProperties(Map<String, String> extraProperties)
```
I checked the usage of `DistributedQueryRunner` and there seems to be only one case (the one you linked) where a single property is set for both extra and coordinator properties. However, it's much more common in `TpchQueryRunner`, so I added these methods there.

AlekseiS commented:
@kokosing thanks for reviewing the changes! I made several changes and commented on the rest.

kokosing commented:
Can you please rebase

AlekseiS commented:
Rebased again.

kokosing commented on presto-main/src/main/java/com/facebook/presto/server/protocol/QueryV1.java:
```diff
@@ -0,0 +1,235 @@
+import static io.airlift.units.DataSize.Unit.MEGABYTE;
+import static java.util.Objects.requireNonNull;
+
+class QueryV1
```
I think you could extract this class in separate commit.

AlekseiS replied:
Doing it in a separate commit won't be easy as the structure of a `Query` is changed and the changes are closely related to introduction of v2 protocol and changes in `StatementResource`.

AlekseiS commented on presto-main/src/main/java/com/facebook/presto/execution/SqlTask.java:
```diff
@@ -78,6 +83,8 @@
     private final AtomicReference<TaskHolder> taskHolderReference = new AtomicReference<>(new TaskHolder());
     private final AtomicBoolean needsPlan = new AtomicBoolean(true);
 
+    private final AtomicReference<Optional<TaskClientOutputContext>> clientOutputContext = new AtomicReference<>(Optional.empty());
```
Getter is not synchronized, so the field needs to be marked as "volatile" for an update to be seen by another thread. Throughout Presto codebase I see atomic primitives are commonly used in such cases instead of volatile.

AlekseiS commented on presto-main/src/main/java/com/facebook/presto/execution/SqlTask.java:
```diff
@@ -333,6 +345,15 @@ public TaskInfo updateTask(Session session, Optional<PlanFragment> fragment, Lis
                     checkState(fragment.isPresent(), "fragment must be present");
+                    // this block must happen before creating task execution to prevent a race condition when execution finished,
+                    // but types are not ready. Types are accessed by a task resource in order to convert pages to a client output
+                    if (fragment.get().getRoot() instanceof OutputNode) {
```
My understanding is that this whole synchronized block happens only once per task because it creates task execution. However, I added a check as you suggested.

AlekseiS commented on presto-client/src/main/java/com/facebook/presto/client/QueryActions.java:
```diff
@@ -0,0 +1,165 @@
+import static com.google.common.base.MoreObjects.toStringHelper;
+
+@Immutable
+public class QueryActions
```
done

AlekseiS commented on presto-main/src/main/java/com/facebook/presto/server/TaskClientOutputContext.java:
```diff
@@ -0,0 +1,52 @@
+
+import static java.util.Objects.requireNonNull;
+
+public class TaskClientOutputContext
```
Done

AlekseiS commented on presto-tests/src/main/java/com/facebook/presto/tests/DistributedQueryRunner.java:
```diff
@@ -412,6 +402,11 @@ public final void close()
         }
     }
 
+    public static Builder builder(Session defaultSession, int workersCount)
```
Done

AlekseiS commented on presto-tests/src/main/java/com/facebook/presto/tests/DistributedQueryRunner.java:
```diff
@@ -91,10 +92,10 @@ public DistributedQueryRunner(Session defaultSession, int workersCount)
     public DistributedQueryRunner(Session defaultSession, int workersCount, Map<String, String> extraProperties)
```
Done

AlekseiS commented:
Rebased and addressed some review comments. Remaining items from review:
* deduplicate parsing logic in `QueryRequestSessionContext`
* think about extending `DistributedQueryRunner.Builder` in `TpchQueryRunner`
* add v2 protocol to product tests

@kokosing , regarding migration, as the first step both protocols can be enabled at the same time. The change allows them to co-exist. During this time some clients/clusters can be switched to using it. Once there's enough confidence, "v2" can be set as the default protocol version in command line and jdbc (once implemented). Later "v1" can be disabled.

BTW, I'm ok merging this PR in parts if you feel more comfortable with it.

kokosing commented:
> BTW, I'm ok merging this PR in parts if you feel more comfortable with it.

v2 is behind the feature flag, so it should be quite safe to merge it as it. We would need just make sure that there is no regression introduced to v1. Anyway, I think it is desired to have someone from FB to take a look at it as well, especially someone well familiar with this code.

AlekseiS commented:
Rebased on 0.193.

AlekseiS commented:
Rebased again to resolve conflicts.

AlekseiS commented:
Deduplicated parsing logic in `QueryRequestSessionContext`.

findepi commented:
> v2 is behind the feature flag

Is it possible to run v1 and v2 side-by-side? That would make the change easier to integrate.

AlekseiS commented:
@findepi , when enabled, v2 will run side by side with v1 on the server.

kokosing commented:
> when enabled, v2 will run side by side with v1 on the server.

But why it is not a default behaviour ? Does v2 affects somehow queries which are using v1?

AlekseiS commented:
> why it is not a default behaviour

@kokosing , I think the main reason is that the protocol is not finished yet, so there might be more changes to it. Other than that, there's no problem enabling it side-by-side with v1.

AlekseiS commented:
Rebased.

AlekseiS commented:
Add a new commit to send dataUris in a cumulative way as they become available.

electrum commented on presto-client/src/main/java/com/facebook/presto/client/ProtocolVersion.java:
```diff
@@ -0,0 +1,34 @@
+
+    public static ProtocolVersion fromString(String value)
+    {
+        requireNonNull(value, "value is null");
```
This can be
```java
return valueOf(value.toUpperCase(ENGLISH));

AlekseiS replied:
Done

AlekseiS commented:
Rebased on (almost) 0.195.
Squashed last commit changing how dataUris are sent, updated wiki.

AlekseiS commented:
Extend `DistributedQueryRunner.Builder` in `TpchQueryRunnerBuilder`.
@kokosing see if the corresponding two commits look better now.

kokosing commented:
@AlekseiS Thank you.

electrum commented on presto-tests/src/test/java/com/facebook/presto/execution/TestQueues.java:
```diff
@@ -284,7 +284,7 @@ public void testClientTagsBasedSelection()
             throws Exception
     {
-        try (DistributedQueryRunner queryRunner = TpchQueryRunner.createQueryRunner(ImmutableMap.of(), ImmutableMap.of("experimental.resource-groups-enabled", "true"))) {
+        try (DistributedQueryRunner queryRunner = TpchQueryRunnerBuilder.builder().setSingleCoordinatorProperty("experimental.resource-groups-enabled", "true").build()) {
```
Wrap here

electrum commented on presto-tests/src/test/java/com/facebook/presto/tests/tpch/TpchQueryRunner.java:
```diff
@@ -1,108 +0,0 @@
-
-import static com.facebook.presto.testing.TestingSession.testSessionBuilder;
-
-public final class TpchQueryRunner
```
Let's keep the class around with just the main method
